File: test/extensions/dynamic_modules/test_data/c/stat_sink_integration_test.c
#include <stddef.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "source/extensions/dynamic_modules/abi/abi.h"

// A stats sink test module whose lifecycle and event hooks emit observable log
// messages via envoy_dynamic_module_callback_log. Integration tests grep Envoy's
// log output for these markers to verify the sink was loaded, configured,
// flushed at least once, and received histogram observations.

static void log_info(const char* message) {
  envoy_dynamic_module_type_module_buffer buf;
  buf.ptr = message;
  buf.length = strlen(message);
  envoy_dynamic_module_callback_log(envoy_dynamic_module_type_log_level_Info, buf);
}

envoy_dynamic_module_type_abi_version_module_ptr envoy_dynamic_module_on_program_init(void) {
  return envoy_dynamic_modules_abi_version;
}

envoy_dynamic_module_type_stat_sink_config_module_ptr
envoy_dynamic_module_on_stat_sink_config_new(
    envoy_dynamic_module_type_stat_sink_config_envoy_ptr config_envoy_ptr,
    envoy_dynamic_module_type_envoy_buffer sink_name,
    envoy_dynamic_module_type_envoy_buffer sink_config) {
  (void)config_envoy_ptr;
  (void)sink_name;
  (void)sink_config;
  log_info("stat sink integration test: config_new called");
  static int config_dummy = 0;
  return &config_dummy;
}

void envoy_dynamic_module_on_stat_sink_config_destroy(
    envoy_dynamic_module_type_stat_sink_config_module_ptr config_module_ptr) {
  (void)config_module_ptr;
  log_info("stat sink integration test: config_destroy called");
}

void envoy_dynamic_module_on_stat_sink_flush(
    envoy_dynamic_module_type_stat_sink_config_module_ptr config_module_ptr,
    envoy_dynamic_module_type_stat_sink_snapshot_envoy_ptr snapshot_envoy_ptr) {
  (void)config_module_ptr;

  // Exercise the snapshot read-back callbacks to prove they are usable during
  // flush. The specific counts depend on the test setup, so we just log them.
  const size_t counter_count =
      envoy_dynamic_module_callback_stat_sink_snapshot_get_counter_count(snapshot_envoy_ptr);
  const size_t gauge_count =
      envoy_dynamic_module_callback_stat_sink_snapshot_get_gauge_count(snapshot_envoy_ptr);

  // Decode each name straight into a stack buffer to exercise the module-provided
  // buffer API. Names longer than the buffer are simply truncated here.
  char name_buf[256];
  size_t name_size = 0;
  uint64_t value = 0;
  uint64_t delta = 0;
  for (size_t i = 0; i < counter_count; i++) {
    envoy_dynamic_module_callback_stat_sink_snapshot_get_counter(
        snapshot_envoy_ptr, i, name_buf, sizeof(name_buf), &name_size, &value, &delta);
  }

  // Look for the always-present "server.uptime" gauge to prove a name round-trips through the buffer
  // API. A small stack buffer is tried first, then a heap retry when the name does not fit, which
  // exercises the truncation-and-retry contract.
  static const char kUptime[] = "server.uptime";
  const size_t uptime_len = sizeof(kUptime) - 1;
  int found_uptime = 0;
  for (size_t i = 0; i < gauge_count; i++) {
    char small[8];
    size_t needed = 0;
    if (!envoy_dynamic_module_callback_stat_sink_snapshot_get_gauge(
            snapshot_envoy_ptr, i, small, sizeof(small), &needed, &value)) {
      continue;
    }
    const char* full = small;
    char* heap = NULL;
    if (needed > sizeof(small)) {
      heap = (char*)malloc(needed);
      if (heap == NULL) {
        continue;
      }
      size_t heap_size = 0;
      envoy_dynamic_module_callback_stat_sink_snapshot_get_gauge(snapshot_envoy_ptr, i, heap, needed,
                                                                 &heap_size, &value);
      full = heap;
    }
    if (needed == uptime_len && memcmp(full, kUptime, uptime_len) == 0) {
      found_uptime = 1;
    }
    free(heap); // free(NULL) is a no-op.
  }

  // Read back any text readouts to exercise the dual-buffer (name and value) getter during flush.
  const size_t text_readout_count =
      envoy_dynamic_module_callback_stat_sink_snapshot_get_text_readout_count(snapshot_envoy_ptr);
  char tr_name[256];
  char tr_value[256];
  size_t tr_name_size = 0;
  size_t tr_value_size = 0;
  for (size_t i = 0; i < text_readout_count; i++) {
    envoy_dynamic_module_callback_stat_sink_snapshot_get_text_readout(
        snapshot_envoy_ptr, i, tr_name, sizeof(tr_name), &tr_name_size, tr_value, sizeof(tr_value),
        &tr_value_size);
  }

  if (found_uptime) {
    log_info("stat sink integration test: found gauge server.uptime");
  }

  char buf[128];
  int n = snprintf(buf, sizeof(buf),
                   "stat sink integration test: flush called counters=%zu gauges=%zu",
                   counter_count, gauge_count);
  if (n < 0 || (size_t)n >= sizeof(buf)) {
    log_info("stat sink integration test: flush called");
    return;
  }
  envoy_dynamic_module_type_module_buffer log_buf;
  log_buf.ptr = buf;
  log_buf.length = (size_t)n;
  envoy_dynamic_module_callback_log(envoy_dynamic_module_type_log_level_Info, log_buf);
}

void envoy_dynamic_module_on_stat_sink_on_histogram_complete(
    envoy_dynamic_module_type_stat_sink_config_module_ptr config_module_ptr,
    envoy_dynamic_module_type_envoy_buffer histogram_name, uint64_t value) {
  (void)config_module_ptr;
  (void)value;
  // The integration test matches on the prefix plus the histogram name passed
  // through from Envoy.
  char buf[256];
  int name_len = (int)histogram_name.length;
  if (name_len > 200) {
    name_len = 200;
  }
  int n = snprintf(buf, sizeof(buf),
                   "stat sink integration test: histogram complete: %.*s",
                   name_len, histogram_name.ptr);
  if (n < 0 || (size_t)n >= sizeof(buf)) {
    log_info("stat sink integration test: histogram complete");
    return;
  }
  envoy_dynamic_module_type_module_buffer log_buf;
  log_buf.ptr = buf;
  log_buf.length = (size_t)n;
  envoy_dynamic_module_callback_log(envoy_dynamic_module_type_log_level_Info, log_buf);
}

File: test/extensions/dynamic_modules/test_data/rust/stat_sink_integration_test.rs
//! Integration test module for stats sink dynamic modules.
//!
//! The lifecycle and event hooks emit log markers via `envoy_log_info!`. The C++ integration test
//! greps Envoy's log output for these markers to verify the sink was loaded, configured, flushed at
//! least once, and received histogram observations. A separate test reads the published gauge value
//! directly to verify the off-main-thread aggregation round trip.

use envoy_proxy_dynamic_modules_rust_sdk::stats_sink::*;
use envoy_proxy_dynamic_modules_rust_sdk::*;
use std::sync::mpsc::{channel, Sender};
use std::sync::Mutex;
use std::thread::JoinHandle;

declare_stat_sink_init_functions!(init, new_stat_sink);

fn init() -> bool {
  true
}

fn new_stat_sink(
  _name: &str,
  _config: &[u8],
  envoy_config: &mut EnvoyStatSinkConfig,
) -> Option<Box<dyn StatSink>> {
  envoy_log_info!("stat sink integration test: config_new called");

  // Define a gauge to publish the aggregated result into, and a scheduler to post that result back
  // to the main thread once it has been computed off the main thread.
  let gauge_id = envoy_config
    .define_gauge("integration_aggregated_counters")
    .expect("gauge definition must succeed during config creation");
  let scheduler = envoy_config.new_config_scheduler();

  // Aggregate snapshots on a dedicated worker thread to model a module that offloads work from the
  // main thread, for example onto a Tokio runtime. The worker owns the scheduler and commits the
  // aggregated total as the event id, which the main thread then publishes into the gauge.
  let (sender, receiver) = channel::<OwnedMetricSnapshot>();
  let worker = std::thread::spawn(move || {
    while let Ok(snapshot) = receiver.recv() {
      let total: u64 = snapshot.counters.iter().map(|counter| counter.value).sum();
      scheduler.commit(total);
    }
  });

  Some(Box::new(TestStatSink {
    gauge_id,
    sender: Mutex::new(Some(sender)),
    worker: Mutex::new(Some(worker)),
  }))
}

struct TestStatSink {
  gauge_id: EnvoyGaugeId,
  sender: Mutex<Option<Sender<OwnedMetricSnapshot>>>,
  worker: Mutex<Option<JoinHandle<()>>>,
}

impl StatSink for TestStatSink {
  fn on_flush(&self, snapshot: &MetricSnapshot<'_>) {
    // Exercise the buffer-based snapshot callbacks to prove names decode straight into a
    // module-owned buffer during flush. A single buffer is reused across every entry, starting
    // empty so the first decode exercises the SDK grow-and-retry path. This is the allocation-free
    // pattern the API enables (for example writing each name to a socket).
    let mut name = Vec::new();
    let mut value = Vec::new();
    let mut tag_name = Vec::new();
    let mut tag_value = Vec::new();
    // Exercise the counter tag callbacks against every counter in the live snapshot. Every tag
    // index the count reports must resolve; a mismatch means the count and the per-tag reader
    // disagree.
    let mut counter_tags_consistent = true;
    for index in 0..snapshot.counter_count() {
      let _ = snapshot.counter(index, &mut name);
      if !snapshot.counter_tag_extracted_name(index, &mut name) {
        counter_tags_consistent = false;
        continue;
      }
      let Some(tag_count) = snapshot.counter_tag_count(index) else {
        counter_tags_consistent = false;
        continue;
      };
      for tag_index in 0..tag_count {
        if !snapshot.counter_tag(index, tag_index, &mut tag_name, &mut tag_value) {
          counter_tags_consistent = false;
        }
      }
    }
    // Decode every gauge name and look for the always-present "server.uptime" gauge, which proves a
    // name round-trips byte-for-byte through the buffer API end to end.
    let mut found_uptime = false;
    // Reconstruct the dimensional name of a tagged gauge from its tag-extracted name and tags, the
    // way a Prometheus-style sink would. "cluster.cluster_0.membership_total" carries a single
    // "envoy.cluster_name" tag whose value is the cluster name, so the tag callbacks must yield
    // tag-extracted name "cluster.membership_total" and tag ("envoy.cluster_name", "cluster_0").
    let mut found_tagged_gauge = false;
    for index in 0..snapshot.gauge_count() {
      if snapshot.gauge(index, &mut name).is_some() && name.as_slice() == b"server.uptime" {
        found_uptime = true;
      }
      if snapshot.gauge_tag_extracted_name(index, &mut name)
        && name.as_slice() == b"cluster.membership_total"
        && snapshot.gauge_tag_count(index) == Some(1)
        && snapshot.gauge_tag(index, 0, &mut tag_name, &mut tag_value)
        && tag_name.as_slice() == b"envoy.cluster_name"
        && tag_value.as_slice() == b"cluster_0"
      {
        found_tagged_gauge = true;
      }
    }
    for index in 0..snapshot.text_readout_count() {
      let _ = snapshot.text_readout(index, &mut name, &mut value);
    }
    if found_uptime {
      envoy_log_info!("stat sink integration test: found gauge server.uptime");
    }
    // A tagged gauge reconstructed from the tag callbacks, and every counter's tag count agreed
    // with its per-tag reads: the tag ABI round-trips end to end against a live snapshot.
    if found_tagged_gauge && counter_tags_consistent {
      envoy_log_info!("stat sink integration test: reconstructed tagged gauge cluster.membership_total envoy.cluster_name=cluster_0");
    }
    envoy_log_info!(
      "stat sink integration test: flush called counters={} gauges={}",
      snapshot.counter_count(),
      snapshot.gauge_count()
    );
    // Copy the snapshot so it outlives this call, then hand it to the worker thread to aggregate.
    if let Some(sender) = self.sender.lock().unwrap().as_ref() {
      let _ = sender.send(snapshot.to_owned());
    }
  }

  fn on_histogram_complete(&self, name: EnvoyBuffer<'_>, _value: u64) {
    envoy_log_info!(
      "stat sink integration test: histogram complete: {}",
      String::from_utf8_lossy(name.as_slice())
    );
  }

  fn on_config_scheduled(&self, envoy_config: &mut EnvoyStatSinkConfig, event_id: u64) {
    // Runs on the main thread. Publish the value aggregated off the main thread into the gauge.
    let _ = envoy_config.set_gauge(self.gauge_id, event_id);
    envoy_log_info!(
      "stat sink integration test: scheduled publish event_id={}",
      event_id
    );
  }
}

impl Drop for TestStatSink {
  fn drop(&mut self) {
    // Close the channel so the worker thread exits, then wait for it to finish.
    self.sender.lock().unwrap().take();
    if let Some(worker) = self.worker.lock().unwrap().take() {
      let _ = worker.join();
    }
    envoy_log_info!("stat sink integration test: config_destroy called");
  }
}
